### ISSUE TITLE
Hash partitioning as a conjunction of hash functions

### ISSUE THREAD

**Issue by bmyerz** [uwescience/raco#473]
Currently we represent hash partitioning as a set of attributes `P <= S`, which represents the partitioning `h(P)`. Sometimes we know more than this. The best example is an equi-join, where we actually know that two disjoint sets of attributes are both partitioned.

e.g.,
```
R = (a,b,c) partitioned on (a, b)
S = (d,e,f) partitioned on (d, e)
T = select * from R join S on a=d and b=e
```

We know that in T, tuples sharing a `(a, b)` are co-located and tuples sharing a `(d, e)` are co-located.
After #467, we only support that T knows `(a, b)` are co-located.

**Comment by senderista** [comment]
I must be missing something, because I don't understand how the logic you added to `ShuffleBeforeJoin` doesn't cover this scenario.

**Comment by bmyerz** [comment]
Maybe this is more clear: I am considering this query

```
s = scan(S);   // schema: a,b,c
t = scan(T);    // schema: a,b,c
r = select * from s, t where s.a=t.a;
store(r, OUTPUT);
```
I want hash_partitioning to provide multiple hash groupings. In this case `MyriaStore.partitioning().hash_partitioned` to return `set(set($0), set($3))`. With #467, we allow only a single hash grouping, so it returns the more conservative `set($0)`.

**Comment by senderista** [comment]
OK, I didn't understand there was an implicit store after the join and you wanted to persist the full partitioning info to the catalog. That makes sense. I'm still not totally clear on the meaning of `set(set($0), set($3))`. Is it supposed to convey that each group in the set can be used equivalently to compute the partitioning of the relation (so we could say that `r` is partitioned equivalently on `s.a` or `t.a`)? How would this be used in a rule (I assume you had some particular use case in mind)?

**Comment by senderista** [comment]
Actually it's pretty obvious how this would be used in rules, never mind (e.g., you could test if a projection or grouping list was a superset of one of the groupings, or if a join condition list was identical to one of the groupings, etc., etc.).

**Comment by bmyerz** [comment]
Yes, more information will allow the plan above the join to remove shuffles.
